### PR TITLE
Feat/usermanager

### PR DESCRIPTION
**Pull request by mswertz**
Added rudimentary user manager

- [x] extended Database interface with getUsers and countUsers, only working when admin
- [x] extended GraphqlAPi with {_admin{users{username},userCount}}, only when logged in as admin 
- [x] when opening EMX2 on /apps/central and sign in as admin you will see 'Admin' menu item
- [x] this Admin user interface has user list and enables create/change username+password
- [x] as a side effect, also draft of ValidationTest has been added, as prep for integration molgenis-expression libary 

Future items might include: search of users; delete of users (big discussion if we want this); firstname/lastname properties.

N.B. a lot of changes are because of pull request 242 that when merged first will remove many changes. Next time I will first checkout blessed master and branch from there ;-)

### CODE REVIEW

**Comment by connoratrug** [comment]
[User manager](https://trello.com/c/HhdDY203/96-user-manager)

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.molgenis.emx2;
+
+public class User {
```
Comment (connoratrug) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: maybe take a look at 'java records' for these types op value classes

Comment (connoratrug) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: what about a user without a usename ?

Comment (connoratrug) [inline backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: without equals user with username 'jan' != 'jan'

Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: at the moment all users HAVE to have user name. Later on, with oicd, we can revisit this.

Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: yes, I was a bit in doubt wether to go that far. So I didn't. You think we should?

Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: Ah yes, that is lovely. `public record User(String username) {}`
So far I had not looked at java 15 features that much. Do we want to go this way now?

File: apps/central/src/components/Admin.vue [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+          label="username"
+          v-model="username"
+          placeholder="Enter username"
+        />
+        <InputPassword
+          label="password"
+          v-model="password"
+          placeholder="Enter password"
+        />
+        <ButtonAction @click="alterUser">Update user</ButtonAction>
```
Comment (connoratrug) [inline apps/central/src/components/Admin.vue:22]: When i create a new user the user is not added to the list until after page reload

**Review by connoratrug — COMMENTED** [review]
fields do not align 
Screenshot 2021-06-07 at 16 16 48

**Review by connoratrug — CHANGES_REQUESTED** [review]
Unclear if 'Users' are the same thing as 'Members' , some pages ask for email while others ask for username. The feature should be described in the docs / help

**Comment by mswertz** [comment]
> Unclear if 'Users' are the same thing as 'Members' , some pages ask for email while others ask for username. The feature should be described in the docs / help

Check. Doesn't indeed mean the same. 'Users' are people who can sign in. 'Members' are Users having a Role within a Database (or 'Schema'). Good suggestion to add this to docs so we start growing that.

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java (hunk 1 above)
Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: For now I add equals + hashcode

Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: Added User(String username) constructor and assert username != null

Comment (mswertz) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:3]: added equals and hashcode

**Comment by mswertz** [comment]
I will add the documentation in a separate pull request otherwise this becomes rather large.

File: backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqAdminFields.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  private static GraphQL grapql;
+  private static Database database;
+  private static final String schemaName = "TestGraphqlAdminFields";
+
+  @BeforeClass
+  public static void setup() {
+    database = TestDatabaseFactory.getTestDatabase();
+    //    PetStoreExample.create(schema.getMetadata());
+    //    PetStoreExample.populate(schema);
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqAdminFields.java:28]: remove if not needed

File: backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqAdminFields.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Test
+  public void testUsers() throws IOException {
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqAdminFields.java:32]: use descriptive name for test

File: backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqAdminFields.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  private JsonNode execute(String query) throws IOException {
+    String result = convertExecutionResultToJson(grapql.execute(query));
+    JsonNode node = new ObjectMapper().readTree(result);
+    if (node.get("errors") != null) {
+      throw new MolgenisException(node.get("errors").get(0).get("message").asText());
+    }
+    return new ObjectMapper().readTree(result).get("data");
+  }
```
Comment (connoratrug) [inline backend/molgenis-emx2-graphql/src/test/java/org/molgenis/emx2/graphql/TestGraphqAdminFields.java:64]: move to helper of parent maybe ?

File: backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/MetadataUtils.java [hunk 6]
```diff
@@ -308,6 +309,25 @@ protected static void alterTableName(DSLContext jooq, TableMetadata table, Strin
...
+    for (Object username :
+        jooq.select(USER_NAME)
+            .from(USERS_METADATA)
+            .orderBy(USER_NAME)
+            .limit(limit)
+            .offset(offset)
+            .fetch()
+            .getValues(USER_NAME)) {
+      users.add(new User((String) username));
+    }
```
Comment (connoratrug) [inline backend/molgenis-emx2-sql/src/main/java/org/molgenis/emx2/sql/MetadataUtils.java:323]: i find these types of loops very hard to read ( and debug) , why not split the fetching and the loop into two statements , of map the result of the fetch

File: backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java (hunk 1 above)
Comment (connoratrug) [reply backend/molgenis-emx2/src/main/java/org/molgenis/emx2/User.java:5]: all of the above and more is solved by records it seems

File: docs/admin.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Admin settings
+
+You can find 'admin' menu when on start page of MOLGENIS, when you view the list of databases (click logo to get there).
+Only when signed in as 'admin' user this menu item will be shown.
+
+## User management
+
+Currently you can use the admin menu to view the users currently registered in the system. In addition you can create a
+new user and/or set a user password.
```
Comment (connoratrug) [inline docs/admin.md:9]: nice 👍 , hope we can keep building on the docs , makes it easer to onboard new projects/devs/users

File: docs/index.html [hunk 8]
```diff
@@ -6,15 +6,29 @@
...
+        padding: 0px;
+    }
+
+    :root {
+        --theme-color: #017FFD;
+        --heading-color: var(--theme-color);
+        --link-color: var(--theme-color);
+        --sidebar-nav-link-color--active: var(--theme-color);
+        --sidebar-nav-color: var(--theme-color);
+        --sidebar-name-font-weight: bold;
```
Comment (connoratrug) [inline docs/index.html:25]: does not work in ie 11 , is that oke ? , should we add a warning ?

**Review by connoratrug — APPROVED** [review]
some minor note, nothing blocking imo , nice to see the docs added.

**Pull request merged**